nemo/collections/nlp/models/language_modeling/megatron/falcon/falcon_decoder_layer.py

```diff
@@ -0,0 +1,284 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Union
+
+import torch
+from megatron.core import parallel_state
+from megatron.core.dist_checkpointing.mapping import ShardedObject, ShardedTensor
+from megatron.core.fusions.fused_bias_dropout import get_bias_dropout_add
+from megatron.core.transformer.attention import SelfAttentionSubmodules
+from megatron.core.transformer.base_layer import BaseLayer, LayerSubmodules
+from megatron.core.transformer.enums import AttnMaskType
+from megatron.core.transformer.identity_op import IdentityFuncOp, IdentityOp
+from megatron.core.transformer.spec_utils import ModuleSpec, build_module
+from megatron.core.transformer.transformer_config import TransformerConfig
+from megatron.core.utils import make_viewless_tensor
+
+""" We use the following notation throughout this file:
+     h: hidden size
+     n: number of attention heads
+     p: number of model parallel partitions
+     np: n/p
+     hp: h/p
+     hn: h/n
+     b: batch size
+     s: sequence length
+     l: number of layers
+    Transformer takes input of size [s, b, h] and returns a
+    tensor of the same size. We use the following arguments:
+        hyperparameters: transformer hyperparameters
+"""
+
+
+@dataclass
+class FalconTransformerLayerSubmodules(LayerSubmodules):
```

Comment:
Why does everything defaults to identity here while these are not in fact based on this code? Could we set the right defaults here right away? https://github.com/NVIDIA/NeMo/blob/1156b6af234a48e366e13c0e8ed9a7c27e9ef678/nemo/collections/nlp/models/language_modeling/megatron/falcon/falcon_spec.py#L34-L57?

Reply:
@janekl FalconTransformerLayerSubmodules are removed in the new refac PR. We will be using https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/main/megatron/core/transformer/transformer_layer.py?ref_type=heads#L19

which initialize all the possible/supported modules as identityop, and later at creation of the layer, init needed modules with whats specified in the spec.